platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/platformstatus/statuslogic/ObservingStatusLogic.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.system.platformstatus.statuslogic;
+
+import com.swirlds.common.system.platformstatus.IllegalPlatformStatusException;
+import com.swirlds.common.system.platformstatus.PlatformStatus;
+import com.swirlds.common.system.platformstatus.PlatformStatusConfig;
+import com.swirlds.common.system.platformstatus.statusactions.CatastrophicFailureAction;
+import com.swirlds.common.system.platformstatus.statusactions.DoneReplayingEventsAction;
+import com.swirlds.common.system.platformstatus.statusactions.FallenBehindAction;
+import com.swirlds.common.system.platformstatus.statusactions.FreezePeriodEnteredAction;
+import com.swirlds.common.system.platformstatus.statusactions.ReconnectCompleteAction;
+import com.swirlds.common.system.platformstatus.statusactions.SelfEventReachedConsensusAction;
+import com.swirlds.common.system.platformstatus.statusactions.StartedReplayingEventsAction;
+import com.swirlds.common.system.platformstatus.statusactions.StateWrittenToDiskAction;
+import com.swirlds.common.system.platformstatus.statusactions.TimeElapsedAction;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+
+/**
+ * Class containing the state machine logic for the {@link PlatformStatus#OBSERVING} status.
+ */
+public class ObservingStatusLogic implements PlatformStatusLogic {
+    /**
+     * The platform status config
+     */
+    private final PlatformStatusConfig config;
+
+    /**
+     * The time at which the platform entered the {@link PlatformStatus#OBSERVING} status
+     */
+    private final Instant statusStartTime;
+
+    /**
+     * The round number of the freeze period if one has been entered, otherwise null
+     */
+    private Long freezeRound = null;
```

Comment:
On the chart, I've described the `OBSERVING` -> `FREEZING` state machine edge like this:

```
transition to FREEZING after an
amount of wall clock time, and
consensus time has advanced
past a freeze timestamp
```

If the status is `OBSERVING` and we receive notification of passing a freeze threshold, the round is saved in the `freezeRound` member variable. Once the observation period has elapsed, we need to have the freeze round to pass into the `Freezing` logic.

If we transitioned to `FREEZING` from `OBSERVING` immediately upon receipt of the freeze notification, the node could begin creating events, and cause a branch

Reply:
Hmm... makes sense. It may be simpler to just transition straight to freeze complete, but at this point in time I'm ok with the current implementation. I think it will become more obvious which is simpler once we get to the point of doing integration.